### PR TITLE
[Batch mode] Cope with bugs that cause error  suppression.

### PR DESCRIPTION
**Pull request by davidungar**
In batch mode, any diagnostics located in non-primary files are suppressed, with the expectation that they will surface when the containing file is compiled as primary. Although this assumption should hold, it does not always. For example, certain inputs can interact with associated type inferencing to violate it.

This PR tracks the current primary input in the `DiagnosticEngine,` which passed it down to the `DiagnosticConsumer.` The `FileSpecificDiagnosticConsumer` uses this information to output the diagnostic into the appropriate primary's `.dia` file. It addresses rdar://49303574.

Needs https://github.com/apple/swift-lldb/pull/1431

### CODE REVIEW

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Review by jrose-apple — DISMISSED** [review]
This can't work as is. We emit diagnostics when we first see a problem, but we might first see a problem for declaration X1 in file X because we're trying to use X1 in file Y. Putting that diagnostic in Y's .dia file isn't correct because it won't always be Y that first tries to use X1.

**Comment by davidungar** [comment]
Please test with following pull request:
https://github.com/apple/swift-lldb/pull/1431

@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
> This can't work as is. We emit diagnostics when we first see a problem, but we might first see a problem for declaration X1 in file X because we're trying to use X1 in file Y. Putting that diagnostic in Y's .dia file isn't correct because it won't always be Y that first tries to use X1.

Thanks! But I don't understand. Isn't this PR doing the same thing as in the pre-batch-mode, single-file case? That's what it seems to do in non-batch mode.

Also, what is the significance of "first"? Why is that important?

File: include/swift/AST/DiagnosticConsumer.h
Comment (jrose-apple) [inline include/swift/AST/DiagnosticConsumer.h]: Hm. I feel really weird having a *DiagnosticConsumer* know about "primary inputs". DiagnosticConsumer is a general interface, and not everything we do has a particular primary file associated with it. Moreover, "primary files" are a Frontend concept, and DiagnosticConsumer is down in AST, so this smells like a conceptual layering violation to me. (Also, I can't remember if whole-module mode uses "primary files" at all.)

An alternate idea I had was to use a SourceLoc as the "current" token, which for now could be the start of the "current" file. I still don't really know what to call it, though.

File: lib/AST/DiagnosticConsumer.cpp
Comment (jrose-apple) [inline lib/AST/DiagnosticConsumer.cpp]: I think you need to remember-for-notes in this case too, so I suggest just pushing this into your new `findSubconsumerAndRememberItForNotes`, or even `subconsumerForLocation`.

File: include/swift/AST/DiagnosticConsumer.h
Comment (davidungar) [reply include/swift/AST/DiagnosticConsumer.h]: Thank you for your clear summary of our talk.

I like your thinking about the layering. If we pass in a `SourceLoc`, we could call it the `diagnosticDestinationOfLastResort`. Or the `finalRestingPlace`. 

OTOH, it might seem unintuitive to use a `SourceLoc` as a stand-in for the `.dia` file. Here's another idea: we could pass in the `.dia` file of the current primary, but then FileSpecificDiagnosticConsumer would need another (unordered) map. What do you think? Any other ideas?

File: lib/AST/DiagnosticConsumer.cpp
Comment (davidungar) [reply lib/AST/DiagnosticConsumer.cpp]: I agree, and I was under the impression that the code in this PR does that. Specifically,`findSubconsumerForPrimaryCausingErrorInNonprimary` returns through `findSubconsumerAndRememberItForNotes`. I see why it was misleading, though I don't have a better alternative.

File: include/swift/AST/DiagnosticConsumer.h
Comment (davidungar) [reply include/swift/AST/DiagnosticConsumer.h]: Another thought, we could call the thing that's passed in to the `DiagnosticEngine`, by it filename or location, the `diagnosticOrphanage`, since it's where homeless ones go. `diagnosticMotel`? (Diagnostics check in but the don't check out.) "Refuge"? I hope all these suggestions will inspire you to think of something really good!

File: lib/AST/DiagnosticConsumer.cpp
Comment (davidungar) [reply lib/AST/DiagnosticConsumer.cpp]: If you think it's clearer, `findSubconsumerForPrimaryCausingErrorInNonprimary` could return a possibly-invalid `SourceLoc`. And the call to `findSubconsumerAndRememberItForNotes` could then be moved up into `handleDiagnostic`. 

Hmm... I have an idea for a better refactoring. Will try it out.

File: include/swift/AST/DiagnosticConsumer.h
Comment (davidungar) [reply include/swift/AST/DiagnosticConsumer.h]: Let me try it with a location and see how it looks

Comment (jrose-apple) [reply include/swift/AST/DiagnosticConsumer.h]: I really want to keep knowledge of FileSpecificDiagnosticConsumer out of DiagnosticEngine as much as possible, mostly to preserve separation of concerns. It already feels weird to me to have it track which file is currently being processed.

`fallbackBufferLocation`? (I put "buffer" in there to make it clear that the exact SourceLoc isn't relevant.) `activeBuffer`? `bufferIndirectlyResponsibleForDiagnostic`?

File: lib/AST/DiagnosticConsumer.cpp
Comment (jrose-apple) [reply lib/AST/DiagnosticConsumer.cpp]: Oh, whoops. I read the code and didn't take that in at all because of the name.

File: lib/AST/DiagnosticConsumer.cpp
Comment (jrose-apple) [inline lib/AST/DiagnosticConsumer.cpp]: I'd just drop the "ForAnyKind" here. This is the general entry point for finding a consumer; the other one specifically calls out the "non-note"-ness. I guess it is edging a little close to the lower-level `subconsumerForLocation`, but the longer name isn't addressing that part anyway.

**Review by jrose-apple — COMMENTED** [review]
Thanks, this factoring of the consumer-finding makes more sense to me too.

File: lib/AST/DiagnosticConsumer.cpp
Comment (davidungar) [reply lib/AST/DiagnosticConsumer.cpp]: Sure! I tried thinking of a name including "AndMakingSureThatNotesWouldBeInTheSamePlaceAsDiagnostics" but I couldn't find anything I liked.

**Comment by davidungar** [comment]
Glad the new factoring works. I have a bug to track down, then I'll rerun the tests.

File: lib/AST/DiagnosticConsumer.cpp
Comment (davidungar) [reply lib/AST/DiagnosticConsumer.cpp]: But your reaction was good because it led to clearer code!

File: include/swift/AST/DiagnosticConsumer.h
Comment (davidungar) [reply include/swift/AST/DiagnosticConsumer.h]: These are good! I'll ponder. Thanks!

Comment (davidungar) [reply include/swift/AST/DiagnosticConsumer.h]: Another design alternative would be to not store the fallbackBufferLocation in the DiagnosticEngine, but for it to merely forward the set method to all consumers. The FileSpecificDiagnosticConsumer would be the only one to store it. How does that alternative feel to you?

**Comment by davidungar** [comment]
The new refactoring led to a more complete change for how we handle non-primaries. I've pushed that so you can see it. Still plan to rename `defaultDiagnosticLoc` to one of your suggestions.

**Comment by davidungar** [comment]
Please test with following pull request:
apple/swift-lldb#1431

@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
Please test with following pull request:
apple/swift-lldb#1431

@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
@jrose-apple I've finished the changes we spoke of, and also they led to a redoing on the subconsumer invariants. (Consumer cannot not be null anymore.) Check it out at your convenience and let me know. (Thanks for the great name!) 

Also, I'm starting to look at the testing aspect. I want to include a test before landing this.

Thanks.

**Comment by davidungar** [comment]
Please test with following pull request:
apple/swift-lldb#1431

@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
Please test with following pull request:
apple/swift-lldb#1431

@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
Please test with following pull request:
apple/swift-lldb#1431

@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
Please test with following pull request:
apple/swift-lldb#1431

@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
Please test with following pull request:
apple/swift-lldb#1431

@swift-ci please smoke test os x platform

File: include/swift/AST/DiagnosticConsumer.h
Comment (jrose-apple) [inline include/swift/AST/DiagnosticConsumer.h]: Can you document the parameter too? It's pretty subtle for someone writing their own DiagnosticConsumer. (It may even be worth mentioning that it's safe to ignore if your consumer doesn't do anything special with different files.)

Comment (jrose-apple) [reply include/swift/AST/DiagnosticConsumer.h]: You could mention `setBufferIndirectlyCausingDiagnosticToInput` too.

File: include/swift/AST/DiagnosticEngine.h
Comment (jrose-apple) [inline include/swift/AST/DiagnosticEngine.h]: Typo: "Diagnost".

File: include/swift/AST/DiagnosticEngine.h
Comment (jrose-apple) [inline include/swift/AST/DiagnosticEngine.h]: Hm, why make the DiagnosticEngine do the string lookup? It feels to me like DiagnosticEngine API ought to traffic in SourceLocs.

File: lib/AST/DiagnosticConsumer.cpp
Comment (jrose-apple) [inline lib/AST/DiagnosticConsumer.cpp]: Nitpick: "fix-its". Also, I don't understand the point here—what makes fix-it consumers different from serializing consumers?

File: lib/FrontendTool/FrontendTool.cpp
Comment (jrose-apple) [inline lib/FrontendTool/FrontendTool.cpp]: Eek. Now it's no longer "for debug info". I don't think this is safe to use because the name for debug info may not be the same as the name of the buffer.

File: lib/FrontendTool/FrontendTool.cpp
Comment (jrose-apple) [inline lib/FrontendTool/FrontendTool.cpp]: This seems overly specific. "This routine gets called in cases where…" seems fine to me.

**Review by jrose-apple — COMMENTED** [review]
Looks reasonable. It needs a new test, though, even if it's a new case in unittests/ that's contrived.

File: lib/FrontendTool/FrontendTool.cpp
Comment (davidungar) [reply lib/FrontendTool/FrontendTool.cpp]: Sure!

File: lib/FrontendTool/FrontendTool.cpp
Comment (davidungar) [reply lib/FrontendTool/FrontendTool.cpp]: Thanks--great catch! I switched it to use the MSF when it's a source file. Does that look right?

File: lib/AST/DiagnosticConsumer.cpp
Comment (davidungar) [reply lib/AST/DiagnosticConsumer.cpp]: Tx for the nitpick. I've clarified the comment, hope it works better now.

File: include/swift/AST/DiagnosticEngine.h
Comment (davidungar) [reply include/swift/AST/DiagnosticEngine.h]: The alternatives that I see are:
1. Expose the `DiagnosticEngine::SourcMgr` instance variable (via a get method?) so that a method in the RAII struct can use it do to the string lookup.
2. Do the lookup in every caller of the RAII class (two, so far).
3. Pass in a SourceManager to the RAII constructor.

Which, if any, are better than what's there now?

File: include/swift/AST/DiagnosticEngine.h
Comment (davidungar) [reply include/swift/AST/DiagnosticEngine.h]: Thanks!! Fixing it.

File: include/swift/AST/DiagnosticEngine.h
Comment (jrose-apple) [reply include/swift/AST/DiagnosticEngine.h]: I was thinking #2 is the best because we probably don't have to do a string lookup in all cases. In particular, SourceFile::getBufferID is available in all cases, right?

(I guess we could make this interface take a buffer ID too, but I don't like that either because it doesn't have a strong type today, and it's also not a type we traffic in in many parts of the compiler. The other advantage of a SourceLoc is that we could some day use it for something more specific.)

File: lib/FrontendTool/FrontendTool.cpp
Comment (jrose-apple) [inline lib/FrontendTool/FrontendTool.cpp]: In case it's useful, here's the idiom for an RAII you may or may not want:

```cxx
Optional<BufferIndirectlyCausingDiagnosticRAII> bicd;
if (condition)
  bicd.emplace(Context.Diags, "foo");
```

File: include/swift/AST/DiagnosticConsumer.h
Comment (davidungar) [reply include/swift/AST/DiagnosticConsumer.h]: Good idea! Done, but a little verbosely.

**Comment by davidungar** [comment]
Regarding testing, see https://github.com/apple/swift/pull/23735/files#diff-43eb268d973af29cc397746d4ed96b7a , which shows a modified test that is part of this PR. The test fails with the status quo and passes with this fix. Is that enough?

File: lib/FrontendTool/FrontendTool.cpp
Comment (davidungar) [reply lib/FrontendTool/FrontendTool.cpp]: Thanks, that's better. I 'll use it.

**Comment by davidungar** [comment]
Please test with following pull request:
apple/swift-lldb#1431

@swift-ci please smoke test os x platform

File: lib/FrontendTool/FrontendTool.cpp
Comment (davidungar) [reply lib/FrontendTool/FrontendTool.cpp]: Tx.

File: include/swift/AST/DiagnosticEngine.h
Comment (davidungar) [reply include/swift/AST/DiagnosticEngine.h]: I tried a couple of alternatives. I quite dislike #2 because we end up with too many lines (for my taste) of duplicated code. Rather, prefer moving the functionality in question to the RAII object, which is sort of a go-between anyway. Also, following your suggestion, the id is obtained without using the filename. So the RAII object gets a `SourceFile`, translates that to a `SourceLoc`, and passes that in to the `DiagnosticEngine`. This layering feels right to me. If you think it best, the RAII definitions and declarations can be moved out of DiagnosticEngine.*. Just tell me where to put them & I'm happy to do it.

File: lib/AST/DiagnosticConsumer.cpp
Comment (jrose-apple) [inline lib/AST/DiagnosticConsumer.cpp]: Typo: "obtains" instead of "occurs".

File: lib/AST/DiagnosticEngine.cpp
Comment (jrose-apple) [inline lib/AST/DiagnosticEngine.cpp]: Nitpick: This won't come up in practice because you *only* modify the setting using RAII, but otherwise: should it return? Or should it reset the buffer instead?

File: test/Misc/serialized-diagnostics-batch-mode-primary-errors.swift
Comment (jrose-apple) [inline test/Misc/serialized-diagnostics-batch-mode-primary-errors.swift]: This test is still relevant, isn't it?

File: test/Misc/serialized-diagnostics-batch-mode-nonprimary-errors.swift
Comment (jrose-apple) [inline test/Misc/serialized-diagnostics-batch-mode-nonprimary-errors.swift]: I can't tell if this one is still relevant or not.

**Review by jrose-apple — COMMENTED** [review]
Implementation looks good. :-) All remaining comments are polish.

**Comment by davidungar** [comment]
Please test with following pull request:
apple/swift-lldb#1431

@swift-ci please smoke test linux platform

File: lib/AST/DiagnosticConsumer.cpp
Comment (davidungar) [reply lib/AST/DiagnosticConsumer.cpp]: I meant "obtains", but "occurs" is better. Fixed, tx.

File: lib/AST/DiagnosticEngine.cpp
Comment (davidungar) [reply lib/AST/DiagnosticEngine.cpp]: Yes, it was reset in an earlier draft. Hmm... This question feels like it illuminates a flaw: We don't account for nested RAII's. Rather than implement a push-down stack for the buffer, let me try just outlawing such nesting for today with an assertion. That will at least detect the occurrence of such a situation.

**Comment by davidungar** [comment]
Please test with following pull request:
apple/swift-lldb#1431

@swift-ci please smoke test os x platform

File: test/Misc/serialized-diagnostics-batch-mode-primary-errors.swift
Comment (davidungar) [reply test/Misc/serialized-diagnostics-batch-mode-primary-errors.swift]: I see, it is partly still relevant because it tests an erroneous primary. I'll modify it and put it back. Thanks for the catch! (Although we may have other tests that cover this.)

File: test/Misc/serialized-diagnostics-batch-mode-nonprimary-errors.swift
Comment (davidungar) [reply test/Misc/serialized-diagnostics-batch-mode-nonprimary-errors.swift]: I don't think it is still needed because it checks for:
1. The absence of "compilation stopped by errors in other files", which is a string which is not in the compiler at all anymore, AFAICT.
2. The empty.dia file existing and being empty, but this same check is in the remaining test, `serialized-diagnostics-batch-mode-nonprimary-errors-only.swift`.

Summary: one predicate is meaningless today because the code is no longer in the compiler, and the other is covered by the remaining test.

**Comment by davidungar** [comment]
Please test with following pull request:
apple/swift-lldb#1431

@swift-ci please smoke test linux platform

File: test/Misc/serialized-diagnostics-batch-mode-nonprimary-errors.swift
Comment (jrose-apple) [reply test/Misc/serialized-diagnostics-batch-mode-nonprimary-errors.swift]: I guess I feel like it's still a slightly different configuration even if it doesn't test any new code paths. But I'll defer to you on this.

Comment (davidungar) [reply test/Misc/serialized-diagnostics-batch-mode-nonprimary-errors.swift]: Thanks. You have a point, but I think I'm confident that this test doesn't carry it's weight.

**Comment by davidungar** [comment]
PS: Thanks for that "Implementation looks good"!

**Review by jrose-apple — APPROVED** [review]

**Comment by davidungar** [comment]
Please test with following pull request:
apple/swift-lldb#1431

@swift-ci please smoke test os x platform

**Pull request merged**
